pil-analyzer/src/type_inference.rs

```diff
@@ -658,6 +659,58 @@ impl TypeChecker {
             })
     }
 
+    /// Type-checks a pattern and adds local variables.
+    fn expect_type_of_pattern(
+        &mut self,
+        expected_type: &Type,
+        pattern: &Pattern,
+    ) -> Result<(), String> {
+        let inferred_type = self.infer_type_of_pattern(pattern)?;
+        self.unifier
+            .unify_types(inferred_type.clone(), expected_type.clone())
+            .map_err(|err| {
+                format!(
+                    "Error checking pattern {pattern}:\nExpected type: {}\nInferred type: {}\n{err}",
+                    self.type_into_substituted(expected_type.clone()),
+                    self.type_into_substituted(inferred_type)
+                )
+            })
+    }
+
+    /// Type-checks a pattern and adds local variables.
+    fn infer_type_of_pattern(&mut self, pattern: &Pattern) -> Result<Type, String> {
+        Ok(match pattern {
+            Pattern::CatchAll => self.new_type_var(),
+            Pattern::Number(_) => {
+                let ty = self.new_type_var();
+                self.unifier.ensure_bound(&ty, "FromLiteral".to_string())?;
+                ty
+            }
+            Pattern::String(_) => Type::String,
+            Pattern::Tuple(items) => Type::Tuple(TupleType {
+                items: items
+                    .iter()
+                    .map(|p| self.infer_type_of_pattern(p))
+                    .collect::<Result<_, _>>()?,
+            }),
+            Pattern::Array(items) => {
+                let item_type = self.new_type_var();
+                for item in items {
+                    self.expect_type_of_pattern(&item_type, item)?;
```

Comment:
Because array has one single base type and tuple has different types for each item.

I mean in the end, `expect` is also not perfect here, because for `[1, y, z]`, we will get the error that `y` is expected to have `FromLiteral` and we ignore what we match again, but I think it's good enough.